Fix dynamic partition selection for postgres planner under join when type of partitioning key different from type of search values

maksm90 opened the pull request:
## Problem description
This fix regards only projection-based implementation of partition selector that works only under postgres planner. The plan has to have the following fragment:
```sql
    Join
     Join Cond: partitioned_table.id = othertable.id
      -> Append
         -> Result
            One-Time Filter: PartSelected
               -> Seq Scan partition1
         -> Result
            One-Time Filter: PartSelected
               -> Seq Scan partition2
         ...
      -> Some materialization point
         -> PartitionSelector for partitioned_table
            Filter: othertable.id
            -> SeqScan othertable
               Filter: <other conditions>
```

Executor under postgres planner performs projection of input tuple to further determine partitions appropriate for dynamic scan node. At the same time, such partition selection mechanics assumes that the type of search tuple after projection has to match to the schema of partitioning table and, as a result, search values inside this tuple must have the same types as in partitioning key. This is so despite of internal comparator used inside partition definition routines allows to compare values with compatible but different types.
Perhaps, this assumption originates from the fact that expressions in the target list of projection have to coerce variables of incoming tuple into values of the same type as at attributes of partitioning key. But there are some data types, e.g. timestamp and date, that are compared directly (there is equality function for them) and aren't cast to the same representation. For these data types planner doesn't build casting functions over attributes and comparator function have to be picked up based on these types. Therefore, the assumption about types of result values after projection equaled to the types of partitioning key is not correct and have to be fixed.

## Query to reproduce
```sql
-- disable ORCA
SET optimizer TO off;

-- create partitioned table with partitioning attribute of timestamp type
CREATE TABLE t_range (a int, b timestamp)
PARTITION BY RANGE (b) (
    START ('2020-06-01') END ('2020-07-31') EVERY (interval '1 month')
);
INSERT INTO t_range
SELECT gen, period
FROM generate_series(1, 1000) gen,
     generate_series('2020-06-01', '2020-07-31', interval '1 month') period;

-- create table for join with date attribute
CREATE TABLE t_inner_to_join (b date);
INSERT INTO t_inner_to_join VALUES ('2020-07-01');

-- perform analyze to build necessary plan
ANALYZE t_range;
ANALYZE t_inner_to_join;

-- query that performs join by timestamp-typed partitioning key and date-typed attribute
-- it have to return 1000 but currently there is returned zero
SELECT COUNT(*) FROM (
    SELECT DISTINCT b
    FROM t_inner_to_join
    WHERE b = '2020-07-01'
) tf, t_range
WHERE t_range.b = tf.b;
```
### Expected result
Plan of last selection has to have the following kind:
```sql
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice3; segments: 3)
         ->  Aggregate
               ->  Nested Loop
                     Join Filter: (t_inner_to_join.b = t_range_1_prt_1.b)
                     ->  Append
                           ->  Result
                                 One-Time Filter: PartSelected
                                 ->  Seq Scan on t_range_1_prt_1
                           ->  Result
                                 One-Time Filter: PartSelected
                                 ->  Seq Scan on t_range_1_prt_2
                     ->  Materialize
                           ->  Partition Selector for t_range (dynamic scan id: 1)
                                 Filter: t_inner_to_join.b
                                 ->  Broadcast Motion 3:3  (slice2; segments: 3)
                                       ->  GroupAggregate
                                             Group Key: t_inner_to_join.b
                                             ->  Sort
                                                   Sort Key: t_inner_to_join.b
                                                   ->  Redistribute Motion 3:3  (slice1; segments: 3)
                                                         Hash Key: t_inner_to_join.b
                                                         ->  GroupAggregate
                                                               Group Key: t_inner_to_join.b
                                                               ->  Seq Scan on t_inner_to_join
                                                                     Filter: (b = '2020-07-01'::date)
 Optimizer: Postgres query optimizer
```
And query has to return non-zero rows.

## Patch description

The current patch remove restriction of projection-based partition selector execution node so that it would be able to apply search values having types different from partitioning key. Also patch contains some refactoring for the second strategy of dynamic partition selection routine based on execution of partitioning level expressions (ORCA case). In particular, it specifies for `tupdesc` argument the role of tuple descriptor for transferable `values` and `isnull` arguments as it's done in the first projection-based strategy. Also it removes `exprTypeOid` argument as unnecessary.

## Affected versions

6X and below

dusx1981 reviewed (approved):
I think there is no problem with this submission

chrishajas commented:
@maksm90  I've reverted this, it introduced a memory leak for inserts into partitioned tables. The
postgres process on the segments goes OOM for sufficiently large
inserts.

Repro:
```
CREATE TABLE rank (id int, rank int, year int, count int)
DISTRIBUTED BY (id)
PARTITION BY RANGE (year)
( START (2006) END (2016) EVERY (1), 
  DEFAULT PARTITION extra ); 
insert into rank select i, 1, 2010, 12 from generate_series(1,100000000)i;
```

maksm90 commented:
> @maksm90 I've reverted this, it introduced a memory leak for inserts into partitioned tables. The
postgres process on the segments goes OOM for sufficiently large
inserts.

Sorry, my bad! I have found buggy place and fixed it. Do I have to post fixed version in a new PR?

chrishajas commented:
Yes please, so that it goes through the PR pipeline again. Thanks for the quick response!
